# MVP

## Joins

### Question 1

Read in all 4 credit card transaction datasets and clean up the column names.

- CardBase.csv
- CustomerBase.csv
- FraudBase.csv
- TransactionBase.csv

In [9]:
import pandas as pd
import numpy as np
from janitor import clean_names

In [3]:
# import required data
cards = pd.read_csv("data/CardBase.csv")
customers = pd.read_csv("data/CustomerBase.csv")
fraud = pd.read_csv("data/FraudBase.csv")
transactions = pd.read_csv("data/TransactionBase.csv")

In [7]:
# Run clean_names on new dataframes
cards.clean_names()
customers.clean_names()
fraud.clean_names()
transactions.clean_names()

transaction_id transaction_date       credit_card_id  transaction_value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     transaction_segment  
0                  SEG25  
1                  SEG16  
2                  SEG14  
3                  SEG17  
4                  SEG14  
...                  ...  
9995               SEG18  
9996               SEG12  
9997               SEG13  
9998               SEG14  
9999               SEG15  

[10000 rows x 5 columns]

### Question 2

Join the data containing card details and customer details by customer id, so that all records of card details and any matching records in customer details are kept.

In [10]:
# View cards dataframe
cards

Card_Number Card_Family  Credit_Limit  Cust_ID
0    8638-5407-3631-8196     Premium        530000  CC67088
1    7106-4239-7093-1515        Gold         18000  CC12076
2    6492-5655-8241-3530     Premium        596000  CC97173
3    2868-5606-5152-5706        Gold         27000  CC55858
4    1438-6906-2509-8219    Platinum        142000  CC90518
..                   ...         ...           ...      ...
495  4708-4407-9601-6022     Premium        691000  CC64993
496  6042-2856-7280-2925        Gold         33000  CC26787
497  8706-3809-5167-3899     Premium        144000  CC32532
498  9540-8558-5897-5046     Premium        830000  CC90246
499  5587-7265-8118-7718        Gold         24000  CC37803

[500 rows x 4 columns]

In [11]:
# View customers dataframe
customers

Cust_ID  Age Customer_Segment Customer_Vintage_Group
0     CC25034   35          Diamond                    VG1
1     CC59625   22          Diamond                    VG1
2     CC69314   33          Diamond                    VG1
3     CC67036   46          Diamond                    VG1
4     CC25597   44          Diamond                    VG1
...       ...  ...              ...                    ...
5669  CC53430   27         Platinum                    VG2
5670  CC31823   33         Platinum                    VG2
5671  CC15335   28         Platinum                    VG2
5672  CC81556   26         Platinum                    VG2
5673  CC38634   21         Platinum                    VG2

[5674 rows x 4 columns]

In [13]:
# Using merge to join dataframes, left join so that cards df keeps its details, joined on Cust_ID
cards.merge(customers, how='left', on='Cust_ID')

Card_Number Card_Family  Credit_Limit  Cust_ID  Age  \
0    8638-5407-3631-8196     Premium        530000  CC67088   27   
1    7106-4239-7093-1515        Gold         18000  CC12076   48   
2    6492-5655-8241-3530     Premium        596000  CC97173   22   
3    2868-5606-5152-5706        Gold         27000  CC55858   30   
4    1438-6906-2509-8219    Platinum        142000  CC90518   50   
..                   ...         ...           ...      ...  ...   
495  4708-4407-9601-6022     Premium        691000  CC64993   34   
496  6042-2856-7280-2925        Gold         33000  CC26787   46   
497  8706-3809-5167-3899     Premium        144000  CC32532   28   
498  9540-8558-5897-5046     Premium        830000  CC90246   35   
499  5587-7265-8118-7718        Gold         24000  CC37803   21   

    Customer_Segment Customer_Vintage_Group  
0            Diamond                    VG1  
1               Gold                    VG3  
2            Diamond                    VG1  
3            Diamond                    VG1  
4           Platinum                    VG2  
..               ...                    ...  
495          Diamond                    VG1  
496         Platinum                    VG2  
497         Platinum                    VG2  
498             Gold                    VG3  
499          Diamond                    VG1  

[500 rows x 7 columns]

### Question 3

Join the data containing fraud details with transaction details so all rows of both tables are kept. What does the resulting row number tell you?

In [14]:
# Viewing df
fraud

Transaction_ID  Fraud_Flag
0     CTID50558449           1
1     CTID55936882           1
2     CTID63762180           1
3     CTID76723439           1
4     CTID21246201           1
..             ...         ...
104   CTID57993591           1
105   CTID91108283           1
106   CTID30494187           1
107   CTID51301522           1
108   CTID49517337           1

[109 rows x 2 columns]

In [15]:
# Viewing df
transactions

Transaction_ID Transaction_Date       Credit_Card_ID  Transaction_Value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     Transaction_Segment  
0                  SEG25  
1                  SEG16  
2                  SEG14  
3                  SEG17  
4                  SEG14  
...                  ...  
9995               SEG18  
9996               SEG12  
9997               SEG13  
9998               SEG14  
9999               SEG15  

[10000 rows x 5 columns]

In [17]:
# Outer merge so both df keep all their rows.
fraud.merge(transactions, how='outer', on='Transaction_ID')

Transaction_ID  Fraud_Flag Transaction_Date       Credit_Card_ID  \
0      CTID50558449         1.0         6-May-16  4246-1369-3659-8804   
1      CTID55936882         1.0        29-Nov-16  1336-9200-1264-2551   
2      CTID63762180         1.0         5-Dec-16  8528-6154-7390-5081   
3      CTID76723439         1.0        15-Sep-16  7908-2695-7391-7499   
4      CTID21246201         1.0        29-Feb-16  2524-4184-5908-6750   
...             ...         ...              ...                  ...   
9995   CTID25037573         NaN        10-Jun-16  9157-2802-8374-1145   
9996   CTID43832298         NaN        23-May-16  7416-4529-6690-5703   
9997   CTID56532072         NaN         6-Aug-16  6699-2639-4522-6219   
9998   CTID88101446         NaN        10-Oct-16  8341-5263-4582-7396   
9999   CTID64730271         NaN        18-Jun-16  1947-8602-1695-7503   

      Transaction_Value Transaction_Segment  
0                  6984               SEG23  
1                 34367               SEG21  
2                 44550               SEG15  
3                 48275               SEG16  
4                 35751               SEG25  
...                 ...                 ...  
9995              44280               SEG18  
9996              27163               SEG12  
9997              36424               SEG13  
9998               8464               SEG14  
9999               7606               SEG15  

[10000 rows x 6 columns]

There are more transactions than fraudulent ones, 1000 to 109. Non fraud transactions are NaN on the merged Fraud_Flag column. 

### Question 4

Join the data containing card details with transaction details so rows from the first which have matching ones in 
the second are returned.

Hint: `semi` is not an allowed option for `how` in `.merge()`. You'll need to think about the process of a semi 
join and then perform each step.

In [21]:
# Using 'inner' as the how as this uses the intersection of the left & Right df's. left_on/right_on used as names are different.
cards.merge(transactions, how='inner', left_on='Card_Number', right_on='Credit_Card_ID')

Card_Number Card_Family  Credit_Limit  Cust_ID Transaction_ID  \
0     8638-5407-3631-8196     Premium        530000  CC67088   CTID85973577   
1     8638-5407-3631-8196     Premium        530000  CC67088   CTID13215166   
2     8638-5407-3631-8196     Premium        530000  CC67088   CTID41016745   
3     8638-5407-3631-8196     Premium        530000  CC67088   CTID82512584   
4     8638-5407-3631-8196     Premium        530000  CC67088   CTID97222954   
...                   ...         ...           ...      ...            ...   
9995  5587-7265-8118-7718        Gold         24000  CC37803   CTID76550379   
9996  5587-7265-8118-7718        Gold         24000  CC37803   CTID58221219   
9997  5587-7265-8118-7718        Gold         24000  CC37803   CTID73252195   
9998  5587-7265-8118-7718        Gold         24000  CC37803   CTID72271256   
9999  5587-7265-8118-7718        Gold         24000  CC37803   CTID81265917   

     Transaction_Date       Credit_Card_ID  Transaction_Value  \
0           13-Feb-16  8638-5407-3631-8196              31592   
1           30-Oct-16  8638-5407-3631-8196               5894   
2            9-May-16  8638-5407-3631-8196              20925   
3           31-Aug-16  8638-5407-3631-8196              36556   
4           25-Nov-16  8638-5407-3631-8196              36715   
...               ...                  ...                ...   
9995        11-Oct-16  5587-7265-8118-7718              23298   
9996        12-Jan-16  5587-7265-8118-7718              49957   
9997         8-Dec-16  5587-7265-8118-7718              13508   
9998         5-Feb-16  5587-7265-8118-7718                624   
9999         5-Apr-16  5587-7265-8118-7718              14019   

     Transaction_Segment  
0                  SEG21  
1                  SEG12  
2                  SEG11  
3                  SEG11  
4                  SEG18  
...                  ...  
9995               SEG15  
9996               SEG18  
9997               SEG25  
9998               SEG24  
9999               SEG13  

[10000 rows x 9 columns]

## Tidy Data

### Question 5

Read in hat_observations and split `observation` into two columns, `hat_colour` and `hat_type`.

Hint: Look up the pandas documentation for `.str.split()`

In [49]:
# import required data & view
hats = pd.read_csv("data/hat_observations.csv")
hats

day  month  year        observation  observation_count
0    25     10  2019        black,beret                  7
1    11      9  2019       blue,top hat                 10
2     4     12  2018    green,santa hat                  9
3    18      6  2018        white,beret                  9
4    26      9  2018    black,santa hat                  1
..  ...    ...   ...                ...                ...
95   15      8  2018  purple,bucket hat                  8
96   26      1  2018   green,bowler hat                  5
97   15     11  2018      black,top hat                  7
98   14      9  2019       green,helmet                  4
99    1     10  2018        blue,fedora                  4

[100 rows x 5 columns]

In [50]:
# Adding two new columns, colour & type, to the existing hats dataframe. Split at the comma.
hats[['hat_colour','hat_type']] = hats.observation.str.split(",", expand = True)

# Dropping old column
hats.drop(columns =["observation"], inplace = True)
 
# View expanded dataframe
hats


day  month  year  observation_count hat_colour    hat_type
0    25     10  2019                  7      black       beret
1    11      9  2019                 10       blue     top hat
2     4     12  2018                  9      green   santa hat
3    18      6  2018                  9      white       beret
4    26      9  2018                  1      black   santa hat
..  ...    ...   ...                ...        ...         ...
95   15      8  2018                  8     purple  bucket hat
96   26      1  2018                  5      green  bowler hat
97   15     11  2018                  7      black     top hat
98   14      9  2019                  4      green      helmet
99    1     10  2018                  4       blue      fedora

[100 rows x 6 columns]

### Question 6

Unite day, month, and year columns into a column called date using a suitable separator. Then find the date where
the most berets were observed.

Hints:
- you can use the `.agg()` function to combine columns
- you may need to convert the data types of the columns you wish to combine

In [54]:
# View data
hats.info

<bound method DataFrame.info of     day  month  year  observation_count hat_colour    hat_type
0    25     10  2019                  7      black       beret
1    11      9  2019                 10       blue     top hat
2     4     12  2018                  9      green   santa hat
3    18      6  2018                  9      white       beret
4    26      9  2018                  1      black   santa hat
..  ...    ...   ...                ...        ...         ...
95   15      8  2018                  8     purple  bucket hat
96   26      1  2018                  5      green  bowler hat
97   15     11  2018                  7      black     top hat
98   14      9  2019                  4      green      helmet
99    1     10  2018                  4       blue      fedora

[100 rows x 6 columns]>

In [55]:
hats["date"] = pd.to_datetime(hats["day"] + "/" + hats["month"] + "/" + hats["year"])
hats

UFuncTypeError: ufunc 'add' did not contain a loop with signature matching types (dtype('<U21'), dtype('<U21')) -> dtype('<U21')